spec outlining packages, namespaces, and imports

jlapacik opened the pull request:
Closes #510 

_Briefly describe your proposed changes:_

This commit specifies how packages, namespaces, and imports work in Flux

_What was the problem?_

Flux currently doesn't have a notion of packages or a way to import external Flux source code.

_What was the solution?_

Propose a design for packages and namespaces as well as the behavior of package imports.

  - [ ] CHANGELOG.md updated with a link to the PR (not the Issue)
  - [ ] Rebased/mergeable
  - [ ] Tests pass
  - [ ] swagger.json updated (if modified Go structs or API)
  - [ ] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -414,7 +414,53 @@ The package clause is not a assignment; the package name does not appear in any
+
+A package defines a namespace via the package statement
+
+    package name
```
Can you define this statement using the EBNF grammar? Then provide and example like this as well.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -414,7 +414,53 @@ The package clause is not a assignment; the package name does not appear in any
+    package name
+
+Here _name_ defines the namespace for all identifiers in the this package.
+Any identifier defined in this package will be referenced via the syntax _name_._identifier_ in all external packages.
```
This is not true, when importing the package the namespace can be renamed in the context of the importing namespace. So this is true only most of the time.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -414,7 +414,53 @@ The package clause is not a assignment; the package name does not appear in any
+Associated with every package is a namespace and a path.
+Namespaces can be imported by another package using that package's path and the import statement as follows:
+
+    import "path/to/package"
```
Same here, can you define the import statement using the EBNF grammar? Then include this example.

nathanielc commented on query/docs/SPEC.md:
````diff
@@ -414,7 +414,53 @@ The package clause is not a assignment; the package name does not appear in any
+```
+import "path/to/package/foo"
+
+variable = foo.x
````
I think this example would be more clear without the assignment.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -414,7 +414,53 @@ The package clause is not a assignment; the package name does not appear in any
+Flux does not support relative imports.
+The _main_ namespace is the only namespace that cannot be imported.
+Assignment cannot be made inside of an imported namespace.
+Namespaces are containers for identifier assignments only.
```
I am not sure what this means.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -414,7 +414,53 @@ The package clause is not a assignment; the package name does not appear in any
+The _main_ namespace is the only namespace that cannot be imported.
+Assignment cannot be made inside of an imported namespace.
+Namespaces are containers for identifier assignments only.
+Because the import statement is not an assignment, a package cannot access nor modify the imported namespaces of its imported namespaces.
```
Lets not state why, but only what. i.e. `A package cannot access nor modify the imported namespaces of its imported namespaces.`

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -414,7 +414,53 @@ The package clause is not a assignment; the package name does not appear in any
+Because the import statement is not an assignment, a package cannot access nor modify the imported namespaces of its imported namespaces.
+An imported namespaces can be renamed via the syntax:
+
+    import new_name_for_foo "path/to/package/foo"
```
Make sure to include this option in the import statement grammar.

pauldix commented on query/docs/SPEC.md:
```diff
@@ -414,7 +414,53 @@ The package clause is not a assignment; the package name does not appear in any
+Any identifier defined in this package will be referenced via the syntax _name_._identifier_ in all external packages.
+
+All identifer assignments in Flux must occur inside of a package.
+The package statement must be the first line of every Flux source file.
```
I assume this isn't true for something that is just a one-off script that is meant to be run as `main`?

jlapacik commented:
I don't think we want to specify how an import path gets resolved just yet. This spec is about namespaces and what their import behavior should be. So I think then a package should be defined as a 2-tuple of (namespace, import path). Where the import path comes from and how it gets resolved will be decided later.

pauldix commented:
@jlapacik ok that makes sense, but that surfaces another question for me. This might only be applicable to imports from supported SCMs (like github), but I was thinking that the import should include a version. Like:

```
import "github.com/pauldix/a", ">=1.0.0"
import "github.com/pauldix/b", "1.1.2"
import "github.com/pauldix/c", "master"
import "github.com/pauldix/d", "ce6195150a7931844acea48d531b212c35be86ed"
```

Might also make sense for imports from specific paths. But then we'd need a version in the package definition, like:

```
package foo "0.1.0"
```

Of course, that puts it on the package developer to increment their version and be responsible. It's how RubyGems did things back in the day.

jlapacik commented:
@pauldix are you saying you want package versions to be required? As in every package statement must have a version associated with it?

pauldix commented:
@jlapacik that was my initial thought. But it's pretty crazy and out there so I'm interested in hearing what others think of it. Best idea or worst idea?

pauldix commented:
Although it would be annoying to have to increment the version in every file in a package. So that would be something that we'd want to be able to declare once.

nathanielc commented:
> Although it would be annoying to have to increment the version in every file in a package. So that would be something that we'd want to be able to declare once.

This is my biggest concern with versions being part of the import statement. That is why typically there is an external file the indicates which versions of the things need to be used. Then that file extends to all imports in the "package", but if a package is a single file it is not that important anymore. 

For now though I want to leave versions out of scope for this PR.  We do not plan to support external imports yet, only builtin packages will be importable. 

When we go to specify how external imports will work, we will need to figure out versions at that point as well.

pauldix commented:
@nathanielc sounds good, let's leave this one for internal packages. Helps so we can start to get those built and move things around. We can have the nerd fight later on versioning and external packages :)

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -573,8 +570,57 @@ Examples:
+If a file does not declare a package statement, all identifiers in that file will belong to the special _main_ package.
+The _main_ package cannot be imported.
+The _main_ package defines the entrypoint of a Flux program.
+That is, only the _main_ package can be executed.
```
What does it mean to be `executed`? I think this needs some more details around side effects etc. Maybe just reference the `main` package here but then in the scopes sections talk about how the `main` package is different.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -573,8 +570,57 @@ Examples:
+The package import path is used by external packages to import and access the identifiers defined in a package.
+
+Access to an identifier assigned in an external package requires referencing the name of the package containing said identifier.
+Access to an identifier assigned in the current package does not require any such reference.
```
These statements feel odd. 

Why do I have to state this?
>Access to an identifier assigned in the current package does not require any such reference.

Shouldn't that be assumed already based on the discussion about identiers and scopes?

> Access to an identifier assigned in an external package requires referencing the name of the package containing said identifier.

This statement is misleading since it won't always be the package name that is used to identify the package when imported.

 Maybe how you access identifiers from within a package should be part of the import statement discussion. Then you can state that access happens by using the package name by default or the specific 'as' name if provided.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -573,8 +570,57 @@ Examples:
+
+Packages can be imported by other packages and their identifiers accessed via the imported package's name.
+This is achieved through the import statement.
+The import statement also allows renaming of the imported package, in which case all identifiers will be reference by the new package name.
```
This sounds great. Let's be careful to not make it sound like we are renaming the package itself. I see the package name and the access identifier as separate things, with the exception that the default access identifier is the package name.

jlapacik commented:
I added a sub-section to the package statement section describing the main package. I also added a tiny section on side effects and which statement produce them.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -572,8 +570,55 @@ Examples:
 
-    Statement = OptionStatement | VarAssignment | ReturnStatement |
-                ExpressionStatement | BlockStatment .
+    Statement = PackageStatement | ImportStatement | OptionStatement |
```
It needs to be the first Statement correct? Then it cannot belong in the Statement production rule as that means it can appear anywhere. You will need to update the Program production rule to have an optional PackageStatement.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -572,8 +570,55 @@ Examples:
+
+A package statement defines a package block.
+Package names must be valid Flux identifiers.
+The package statement must be the first line of every Flux source file.
```
Should we say first `line` or first `statement`? Meaning do we care about extra whitespace at the beginning of the file?

jlapacik replied:
Yes should be first statement.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -572,8 +570,55 @@ Examples:
+
+1. It defines the entrypoint of a Flux program
+2. It cannot be imported
+3. All composite query operations defined in the _main_ package are coerced into producing side effects
```
I like this phrasing `are coerced into producing side effects` what are `composite query operations`? Do we define that anywhere else in the spec? If not lets give a name to functions that produce query operations. Then we can refer to it here.

I think in the section about `Query Spec` we should talk about side effects and give a name to functions that produce a query operation.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -572,8 +570,55 @@ Examples:
+
+A package's import path is always absolute.
+Flux does not support relative imports.
+Assignment cannot be made inside of an imported package.
```
This bit kind of reads like packages are not allowed to have assignment statements. Maybe we should say `Assigment into an imported package identifier is not allowed`?

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -650,6 +695,15 @@ Examples:
 
+### Side Effects
+
+Side effects can occur in two ways in Flux.
```
`Side effects can occur in two ways`.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -650,6 +695,15 @@ Examples:
+1. By reassigning builtin options
+2. By calling a function that produces side effects
+
+In Flux a function produces side effects if it is explicitly declared to have side effects or if it calls a function that itself produces side effects.
```
Remove the `In Flux` bits.

nathanielc replied:
Also ` produces side effects if it` -> ` produces side effects when it`. I like using `when` instead of `if` because it reads more like statement of fact instead of a procedure for how to do something. Just a small nit.

nathanielc commented on query/docs/SPEC.md:
```diff
@@ -441,6 +441,10 @@ In addition to explicit blocks in the source code, there are implicit blocks:
 
+**Note:** The syntax of a _file block_ is defined as:
+
+    FileBlock = [PackageStatement] StatementList
```
I know we refer to the file block here, but lets maybe just call this the Program rule? And the Program rule should look like this:

```
    Program = [PackageStatement] [ImportList] StatementList
```

Where the ImportStatement is not part of the normal sets of statements. Meaning all import must come before any other statements but after the package statement.

jlapacik replied:
Updated.
